Add 202108 New Blog Deep Equilibrium Models (#1342)
JuliaLang/www.julialang.org: blog/2021/08/DEQ.md

is saying is that, if you have a deep neural network with $n$ very large layers, you need to backpropogate
through $n$ neural networks. **But if $n$ is infinite, you only need to backpropogate through 1!** The details
of this have been well-studied in the scientific computing literature since at least the 90's. For example, 
Steven Johnson's [Notes on Adjoint Methods for 18.335 from 2006](https://math.mit.edu/~stevenj/18.336/adjoint.pdf) 
shows a well-written derivation of an adjoint equation ("backpropagation" equation) for a rootfinding solver,
along with a [liteny](https://link.springer.com/content/pdf/10.1007%2F3-540-45718-6_20.pdf) of 
[papers](http://www.jcomputers.us/vol5/jcp0503-11.pdf) that [use](https://ieeexplore.ieee.org/document/4724607) this
[result](https://www.computer.org/csdl/proceedings-article/cis/2008/3508a020/12OmNz61djv) in the 90's and 00's
to mix neural networks and nonlinear solving. We note very briefly that solving for a steady state is equivalent
to solving a system of nonlinear algebraic equations $NN(x) - x = 0$, since finding this solution would give

> $x_{ss} = NN(x_{ss})$ ,or the steady state.

Reviewer: ```suggestion
$x_{ss} = NN(x_{ss})$, or the steady state.
```

But everything in this world is a differential equation, so let's take a turn and twist this into an ODE!

## Mixing DEQs and Neural Ordinary Differential Equations (Neural ODEs)

> From the viewpoint of Julia and the DiffEqFlux.jl library, it is also natural to look at DEQ from

Reviewer: ```suggestion
From the viewpoint of Julia and the DiffEqFlux.jl library, it is also natural to look at DEQs from
```

a differential equations perspective. Instead of viewing the dynamical system as a discrete process 
$x_{n+1} = f(x_n)$, we can equivalently view the system as evolving
continuously, i.e. $x' = f(x)$. If we think about $dx/dt = f(x)$, by Euler's method we approximate $dx = x_{n+1} - x_n$
and simplify to get $x_{n+1} = x_n + dt f(x_n) = g(x_n)$ which relates us back to our original definition with a
slight change to the function. However, in this ODE sense, convergence is when the change is zero, 
or $x'=0$, which again happens when $f(x)=0$ and is a rootfinding problem. **But this view is insightful: 
a DEQ is a neural ODE where time goes to infinity.** Now, instead of taking 1 step at a time, we can take $dt$
steps at a time towards the steady state. This means an adaptive ODE solver can notice we are converging and take
larger and larger steps to get to that equilibrium a bit quicker. But also, given DiffEqFlux, this observation makes implementing
DEQ models in Julia insanely simple. Let's go for it!

> ## Let us implement a simple DEQ Model via ODE Solvers

Reviewer: ```suggestion
## Let us implement a simple DEQ Model via ODE Solvers with Event Handling
```

The Julia [DifferentialEquations.jl](https://diffeq.sciml.ai/stable/) library has a problem type known as
[SteadyStateProblem](https://diffeq.sciml.ai/stable/types/steady_state_types/) which automatically solves
until $x'$ is sufficiently small (below tolerance), in which case it will automatically use a
[terminating callback](https://diffeq.sciml.ai/stable/features/callback_functions/#Example-2:-Terminating-an-Integration)
to exit the iteration at the (approximately) found steady state. Because the [SciML Organization's Packages](https://sciml.ai/)
are differentiable, we can stick neural networks inside of this "steady state of ODEs" problem, and that

> will generate a training mechanism for this continuous-stepping DEQ procedure.

Reviewer: ```suggestion
will generate a training mechanism for this continuous-stepping DEQ procedure. This `SteadyStateProblem`
solution then uses the [Nonlinear Solve Adjoint](https://math.mit.edu/~stevenj/18.336/adjoint.pdf) to calculate
the backpropagation in the efficient manner without requiring backpropagation of the iterations. This thus
gives an efficient implementation of a DEQ without requiring any new tooling or packages, but can also 
outperform the fixed-point iteration approaches by taking multiple steps at a time.
```

The following code block creates a DEQ model. The acute will notice that this code looks
awfully similar to typical [Neural ODEs implemented in Julia](https://julialang.org/blog/2019/01/fluxdiffeq/).
Therefore, the DEQ implementation simply adds an extra steady state layer on top
of the ODE function, and as long as we use the correct sensitivity corresponding to
steady state problems, we are covered.
 
```julia
using Flux
using DiffEqSensitivity
using SteadyStateDiffEq
using DiffEqFlux
using OrdinaryDiffEq
using CUDA
CUDA.allowscalar(false)

ann = Chain(Dense(1, 2), Dense(2, 1)) |> gpu
p,re = Flux.destructure(ann)
tspan = (0.0f0, 1.0f0)

function solve_ss(x)
    xg = gpu(x)
    z = re(p)(xg) |> gpu
    function dudt_(u, _p, t)
        # Solving the equation f(u) - u = du = 0
        re(_p)(u+xg) - u
    end
    ss = SteadyStateProblem(ODEProblem(dudt_, gpu(z), tspan, p))

>     x = solve(ss, DynamicSS(Tsit5()), u0 = z, abstol = 1e-5, reltol = 1e-5).u

Reviewer: ```suggestion
    x = solve(ss, DynamicSS(Tsit5()), u0 = z, abstol = 1f-5, reltol = 1f-5).u
```

and when put together you get a DEQ!**

## Generalizing to other Solution Techniques

There are many ways that one can solve a rootfinding problem with different characteristics.
One can directly use Newton's method, but this can require a good guess and may not distinguish
between stable and unstable equilibrium. Julia packages like 
[NLsolve.jl](https://github.com/JuliaNLSolvers/NLsolve.jl) provide many feasible solutions, and
Bifurcation tools like [BifurcationKit.jl](https://rveltz.github.io/BifurcationKit.jl/dev/) give a whole host of other methods. 
Given the importance of solving nonlinear algebraic systems and their differentiability, the SciML organization
has put together a common interface package [NonlinearSolve.jl](https://nonlinearsolve.sciml.ai/dev/) that weaves
together all of the techniques throughout the package ecosystem (bringing together methods from SUNDIALS, MINPACK,
etc.) and generally defines its differentiability. As such, this package gives a "one-stop shop" for weaving both new
implementations and classical FORTRAN implementations with machine learning without having to worry
about the training details.

Let's see this in action!

> (Demodemodemo)

Reviewer: @QiyaoWei can you try and get this demo together?

continuous for machine learning, and from the explicit to the black-box for differential
equations.

What follows in recent years is a surge of algorithms expanding the frontiers of applying
Neural ODEs to physical problems, as well as studies exploring the theoretical implications
of this new-born area of research. For the former, one can refer to a flurry of works in the
scientific machine learning (SciML) community, many of which based on the flexible structure
of the Julia programming language. Among the latter direction, Deep Equilibrium Models (DEQ)
stand out as an extension of Neural ODE theory, in particular making the notion of infinite-depth
neural networks explicit in the algorithm. This blog post will be aimed at exploring the
properties of DEQ models.

(Note: If you are interested in this work and are an undergraduate or graduate
student, we have [Google Summer of Code projects available in this area](/soc/ideas-page). This
[pays quite well over the summer](https://developers.google.com/open-source/gsoc/help/student-stipends).
Please join the [Julia Slack](http://julialang.org/slack/) and the #jsoc channel to discuss in more detail.)

\toc

> ## What is a Neural Ordinary Differential Equation?

Reviewer: I would just delete this entire section. It's just rambling and doesn't actually describe neural ODEs.

To put it simply, Neural ODE uses a neural network to parameterize what ODE we are
solving. This is exactly what I was referring to when I noted that the discovery of Neural ODE
allowed the differential equations community to shift from using explicit definitions to
black-box solvers. Instead of having to specify the exact position-momentum relations of a
pendulum using Hamiltonian equations, one can train a neural network to implicitly represent
the differential equations that describe the system. In the case where we know
exactly the differential equations from physics, we should rarely expect the neural network
to outperform human knowledge. Nevertheless, this method crowns itself as an implicit

> modelling technique which can work when we only observe input and output data of the system

Reviewer: ```suggestion
modeling technique which can work when we only observe input and output data of the system
```

without possessing knowledge of what happens inside. Then, one of the biggest pros is to
just fit the terms which you don't know exactly. There are a bunch of differential equations
in fluid dynamics for example, which would almost fall under "we know exactly the
differential equations" after many approximations, and in realistic systems some additional terms might pop up as well.

It is well known in literature that Neural ODE can be seen as a continuous version
of ResNets. This is what I was referring to when I noted that Neural ODE caused

> a shift in perspective for the machine learning community from discrete to continuous. We

Reviewer: this sentence has a different narrator.

have always known that powerful techniques such as gradient-based optimization combined with
discrete neural network models gives us the great inference engine that is "standard deep learning".
With the addition of Neural ODE, we can take a look at what happens when we take infinitely
small steps in adjustment (think learning rate $\lim _{\lambda \rightarrow 0}$), effectively resulting in a differential equation.
For our purpose, we will take inspiration from this property of Neural ODE, and

> approximate an "infinitely deep" model as each network layer tends to zero in size.

Reviewer: this whole paragraph doesn't seem useful. It seems like just some rambling.

@def rss_pubdate = Date(2021, 1, 18)
@def rss = """ Implementing Deep Equilibrium Models in Julia... """
@def published = "18 January 2021"

> @def title = "Implementing Deep Equilibrium Models in Julia"

Reviewer: ```suggestion
@def title = "Composability in Julia: Implement Deep Equilibrium Models via Neural ODEs"
```

> @def authors = """Chris Rackauckas, Mike Innes, Yingbo Ma, Jesse Bettencourt, Lyndon White, Vaibhav Dixit"""  

Reviewer: update the authors

<!-- Translations: [Traditional Chinese](/blog/2019/04/fluxdiffeq-zh_tw) -->

In this blog post we will show you how to easily, efficiently, and
robustly use steady state solvers with neural networks in Julia.

<!-- ![Flux ODE Training Animation](https://user-images.githubusercontent.com/1814174/51399500-1f4dd080-1b14-11e9-8c9d-144f93b6eac2.gif) -->

The combination of differential equations and neural networks has aroused wide-spread interest
since 2017. For more on the history of this field, please see the seminal work
[Neural Ordinary Differential Equations](https://arxiv.org/abs/1806.07366) and our previous
blog post [DiffEqFlux.jl – A Julia Library for Neural Differential Equations](https://julialang.org/blog/2019/01/fluxdiffeq/).
In my opinion, this is one of the best, if not

> the most successful, method that interweaves machine learning with another field of research.

Reviewer: this is a bit hyperbolic.

In doing so, not only do we get to reap the benefits of centuries of research in the
differential equations community, we also shift our perspective from the discrete to the

> continuous for machine learning, and from the explicit to the black-box for differential

Reviewer: the differential equations are white boxed!

Reviewer: I see what you're going for from the paragraph later. So I think should be rephrased though I am not sure how right now.

equations.

What follows in recent years is a surge of algorithms expanding the frontiers of applying
NeuralODEs to physical problems, as well as studies exploring the theoretical implications
of this new-born area of research. For the former, one can refer to a flurry of works in the
scientific machine learning (SciML) community, many of which based on the flexible structure
of the Julia programming language. Among the latter direction, Deep Equilibrium Models (DEQ)

> stand out as an extension of NeuralODE theory, in particular making the notion of infinite-depth

Reviewer: NeuralODE  -> Neural ODE

neural networks explicit in the algorithm.

(Note: If you are interested in this work and are an undergraduate or graduate
student, we have [Google Summer of Code projects available in this area](/soc/ideas-page). This
[pays quite well over the summer](https://developers.google.com/open-source/gsoc/help/student-stipends).
Please join the [Julia Slack](http://julialang.org/slack/) and the #jsoc channel to discuss in more detail.)

\toc

## What is a Neural Ordinary Differential Equation?

> To put it simply, NeuralODE simply uses a neural network to parameterize what ODE we are

Reviewer: ```suggestion
To put it simply, NeuralODE uses a neural network to parameterize what ODE we are
```

solving. This is exactly what I was referring to when I noted that the discovery of NeuralODE
allowed the differential equations community to shift from using explicit definitions to

> black-box solvers. Instead of having to specify the exact position-momentum relations of a

Reviewer: black-box universal function definitions.

> pendulum using Hamiltonian equations, one could simply train a neural network to represent

Reviewer: ```suggestion
pendulum using Hamiltonian equations, one can train a neural network to represent
```

> implicitly the differential equations that describe the system. In the case where we know

Reviewer: ```suggestion
the differential equations that describe the system. In the case where we know
```

exactly the differential equations from physics, we should rarely expect the neural network

> to outperform human knowledge. Nevertheless, this method crowns itself as an implicit
> modelling technique, especially when we only observe input and output data of the system

Reviewer: ```suggestion
to outperform human knowledge. Nevertheless, this method crowns itself as an implicit
modelling technique which can work when we only observe input and output data of the system
```

without possessing knowledge of what happens inside.

It is well known in literature that NeuralODE can be seen as a continuous version

> of ResNets. This is exactly what I was referring to when I noted that NeuralODE caused
> a shift in perspective for the machine learning community from discrete to continuous. We

Reviewer: ```suggestion
of ResNets. This is what I was referring to when I noted that NeuralODE caused
a shift in perspective for the machine learning community from discrete to continuous. We
```

> have always known that powerful techniques such as gradient optimization combined with
> discretized computation gives us the great inference engine that is "deep learning".

Reviewer: ```suggestion
have always known that powerful techniques such as gradient-based optimization combined with
discrete models gives us the great inference engine that is "standard deep learning".
```

With the addition of NeuralODE, we can take a look at what happens when we take infinitely
small steps in adjustment, effectively resulting in a differential equation. It is precisely
this property of NeuralODE that allows it to approximate an "infinitely deep" model
as each network layer tends to zero in size.

> As it turns out, there is a line of work following [Deep Equilibrium Models](https://arxiv.org/abs/1909.01377)
> that explores the notion of infinitely-deep neural networks in an explicit manner. It borrows
> much of the ideas from the NeuralODE line of research, and allows neural network structures
> to be viewed as repeated applications of layered computations. This is often very accurate,
> for example when we apply convolution filters on images---the network can essentially
> be viewed as a convolutional feature extractor consisting of repetitive blocks, and one
> linear output layer at the very end. When there is an infinite number of these repeated
> blocks, the network might explode off to infinity. If it doesn't, however, it reaches an
> equilibrium points and converges, hence the name.
> 
> ## DEQ Models and Infinitely Deep Networks

Reviewer: ```suggestion
## Deep Equilibrium Models and Infinitely Deep Networks

There is a line of work following [Deep Equilibrium Models](https://arxiv.org/abs/1909.01377)
that explores direct representations of infinitely-deep neural networks. It borrows
much of the ideas from the NeuralODE line of research, and allows neural network structures
to be viewed as repeated applications of layered computations. This is often very accurate,
for example when we apply convolution filters on images---the network can essentially
be viewed as a convolutional feature extractor consisting of repetitive blocks, and one
linear output layer at the very end. 

However, there are caveats to consider when taking the repetition to the limit. 
When there is an infinite number of these repeated blocks, the value being
pushed forward might continually increase until it explodes to infinity. But under
certain conditions, this repetition `x_n+1 = x_n + f(x_n)` does not diverge but
rather converges to a stable equilibrium. The central idea behind Deep Equilibrium
Models (DEQs) is to capture the space of convergent dynamical systems using a
neural network representation of the residual `f`.
```

Reviewer: need to figure out tex for that

> Let us begin by looking at an instantiation of the equations for DEQ models.

Reviewer: First let's define a multilayer perceptron neural network. ...

$$
\begin{array}{l}
z:=0\\
\text { Repeat until convergence: }\\
z:=\tanh (W z+x)
\end{array}
$$

Which eventually brings us to
$$
z^{\star}=\tanh \left(W z^{\star}+x\right)
$$

The idea of DEQ is, in a sense, insanely simple! We assume there are infinitely many

> compositions of identical layers (equivalently an infinitely deep network), and

Reviewer: ```suggestion
compositions of this kind of layer and
```

> directly solve for the fixed point described by this equation. Clearly, naively iterating

Reviewer: ```suggestion
solve for the fixed point described by the dynamical system (INSERT EXACT FORM HERE). Clearly, naively iterating
```

> directly solve for the fixed point described by this equation. Clearly, naively iterating
> for the fixed point would take an enormous amount of time. Thanks to the Implicit Function
> Theorem, DEQ shows that it is possible to train an infinitely-deep neural network using the
> gradient optimization toolbox. For more reference, please see
> [Deep Implicit Layers](http://implicit-layers-tutorial.org)
> 
> From the viewpoint of Julia and the DiffEqFlux.jl library, it is also natural to look at DEQ from
> a differential equations perspective. Thanks to NeuralODE, the machine learning community is now able
> to put on the glasses of differential equations, and see through a new perspective.
> If we perceive the iterating layer of the network as an ODE, DEQ is solving nothing but the
> well-known "Steady State Problem", defined as the final converging state of the ODE as time
> goes to infinity. This line of work is well-studied in the scientific machine learning
> literature, see for example [Notes on Adjoint Methods for 18.335](https://math.mit.edu/~stevenj/18.336/adjoint.pdf)

Reviewer: ```suggestion
directly solve for the fixed point described by this equation. 

Clearly, naively iterating for the fixed point would take an enormous amount of time. However,
classical dynamical systems theory relates this infinite process to a simple mathematical problem.
If we perceive the iterating layer of the network as a dynamical system, DEQ is solving nothing but the
well-known "Steady State Problem", defined as the final converging state of the system as time
goes to infinity. For an update equation $x_{n+1} = x_n + f(x_n)$, "the equation is not changing anymore
is equivalent to saying $x_{n+1} = x_n$ which reduces the problem to finding the $x$ such that $f(x) = 0$.
This is also known as a rootfinding problem, where the forward and adjoint methods have been well-studied
in the scientific computing literature since at least the 90's. For example, Steven Johnson's 
[Notes on Adjoint Methods for 18.335 from 2006](https://math.mit.edu/~stevenj/18.336/adjoint.pdf) shows a
clean derivation of an adjoint equation ("backpropagation" equation) for a rootfinding solver. 
```

> 
> $$
> f(u) - u = du = 0
> $$
> 
> Reaching a fixed point of the iterative layer application is much like solving for the final
> value in a converging sequence. To put it simply, we set the current value equal to the next,
> and obtain a solution for this recursion problem. As we show above, if we use `u` to represent
> the current layer and `f(u)` to represent the next layer, we can sloppily denote the fixed point
> equation as `f(u)-u=0`. Crucially, this can also be converted into a steady state problem
> by defining an ODE layer `f(u)-u=du` and solving
> that ODE in its steady state. This gives us equality among all three components.

Reviewer: ```suggestion
## Mixing DEQs and Neural ODEs

From the viewpoint of Julia and the DiffEqFlux.jl library, it is also natural to look at DEQ from
a differential equations perspective. Thanks to NeuralODE, the machine learning community is now able
to put on the glasses of differential equations, and see through a new perspective. Instead of viewing the
dynamical system as a discrete process $x_{n+1} = x_n + f(x_n)$, we can equivalently view the system as evolving
continuously, i.e. $x' = f(x)$. In this sense, convergence is when the change is zero, or $x'=0$, which again
happens when $f(x)=0$ and is a rootfinding problem. But this view is insightful: a DEQ is a neural ODE where
time goes to infinity.
```

> ## Let us implement a simple DEQ Model

Reviewer: ```suggestion
## Let us implement a simple DEQ Model via ODE Solvers
```

> 
> The following code block creates a DEQ model. The acute will notice that this code looks

Reviewer: ```suggestion
There are many ways that one can solve a rootfinding problem with different characteristics.
One can directly use Newton's method, but this can require a good guess and may not distinguish
between stable and unstable equilibrium. Bifurcation tools like [BifurcationKit.jl](https://rveltz.github.io/BifurcationKit.jl/dev/) give a whole host of other methods. But for our demonstration,
we will focus on showcasing the evolving ODE approach. This is similar to the fixed point method of running
$x_{n+1} = x_n + f(x_n)$, but we can instead allow the ODE solver to adaptively change its steps along the
trajectory to better facilitate convergence. We can define "convergence" as a tolerance where for $x' = f(x)$,
we want $x'$ to be sufficiently small (in absolute and relative tolerance). It turns out that a method for
doing this, along with its backpropogation, is already defined in Julia as the 
[steady state problem](https://diffeq.sciml.ai/stable/types/steady_state_types/) which will automatically
detect when the ODE has converged and halt the integration. So let's demonstrate a DEQ via SteadyStateProblem.

The following code block creates a DEQ model. The acute will notice that this code looks
```

pharmacometric simulation engine [PuMaS.jl](https://doi.org/10.1007/s10928-018-9606-9)
into the deep learning framework. With access to the full range of solvers for ODEs,
SDEs, DAEs, DDEs, PDEs, discrete stochastic equations, and more, we are
interested to see what kinds of next generation neural networks you will build with Julia.

Note: a citable version of this post is published on [Arxiv](https://arxiv.org/abs/1902.02376).

```
@article{DBLP:journals/corr/abs-1902-02376,
  author    = {Christopher Rackauckas and
               Mike Innes and
               Yingbo Ma and
               Jesse Bettencourt and
               Lyndon White and
               Vaibhav Dixit},
  title     = {DiffEqFlux.jl - {A} Julia Library for Neural Differential Equations},
  journal   = {CoRR},
  volume    = {abs/1902.02376},
  year      = {2019},
  url       = {https://arxiv.org/abs/1902.02376},
  archivePrefix = {arXiv},
  eprint    = {1902.02376},
  timestamp = {Tue, 21 May 2019 18:03:36 +0200},
  biburl    = {https://dblp.org/rec/bib/journals/corr/abs-1902-02376},
  bibsource = {dblp computer science bibliography, https://dblp.org}
}

> ```

Reviewer: needs a new conclusion

<!---

> https://github.com/SciML/DiffEqFlux.jl/issues/567

Reviewer: @DhairyaLGandhi can we get this fixed up on the Flux side?

@def rss_pubdate = Date(2021, 1, 18)
@def rss = """ Implementing Deep Equilibrium Models in Julia... """
@def published = "18 January 2021"
@def title = "Composability in Julia: Implement Deep Equilibrium Models via Neural ODEs"
@def authors = """Chris Rackauckas, Mike Innes, Yingbo Ma, Jesse Bettencourt, Lyndon White, Vaibhav Dixit"""  

<!-- Translations: [Traditional Chinese](/blog/2019/04/fluxdiffeq-zh_tw) -->

In this blog post we will show you how to easily, efficiently, and

> robustly use steady state solvers with neural networks in Julia.

Reviewer 2: it dives very quickly into the topic IMO. Without directly looking into the code at the end, it is not clear what you actually demonstrate in the post. Maybe motivate in the beginning DEQs:

- what are they good for/which tasks can be solved. 
- how can they be described mathematically. (specifically that might be a good point to say it's implicitly defined so you can choose whatever method you want to solve them.. )
- how you would likely solve it with the methods in NLsolve, Bifurcations, etc.
- differences wrt to traditional NN architectures. 

(I guess it's good to make first sure everyone has that information about DEQs)

Then state that you'll implement a DEQ for XYZ problem based on its formulation as a steady state problem of a dynamical system. Highlighting that his needs compositionality between ML and DiffEq tools. Try to make sure that people understand the problem setting well.. 

- make the connection to steady states of ODEs very clear from a mathematical perspective 
- introduce Neural ODEs (at this point it should be rather easy to introduce them. So you can start with a more solid ground and it's clear what a "infinitely small" layer actually is..)
- say what advantages that viewpoint has.

- go to the examples

- for the conclusion: I think it's nice to mention that this basically worked without any new code. (Sure we had to fix some bugs wrt to the type stability of the Steady state solver or the CUDA issue in the adjoints but at least in principle it just requires tools which were already there.) 
In the same spirit, I think writing here many things more like: "this is actually nothing else than a steady state/non-linear solve on top of a neural network" with some more equations would be great.

In doing so, not only do we get to reap the benefits of centuries of research in the
differential equations community, we also shift our perspective from the discrete to the
continuous for machine learning, and from the explicit to the black-box for differential
equations.

What follows in recent years is a surge of algorithms expanding the frontiers of applying
NeuralODEs to physical problems, as well as studies exploring the theoretical implications
of this new-born area of research. For the former, one can refer to a flurry of works in the
scientific machine learning (SciML) community, many of which based on the flexible structure
of the Julia programming language. Among the latter direction, Deep Equilibrium Models (DEQ)
stand out as an extension of NeuralODE theory, in particular making the notion of infinite-depth
neural networks explicit in the algorithm.

(Note: If you are interested in this work and are an undergraduate or graduate
student, we have [Google Summer of Code projects available in this area](/soc/ideas-page). This
[pays quite well over the summer](https://developers.google.com/open-source/gsoc/help/student-stipends).

> Please join the [Julia Slack](http://julialang.org/slack/) and the #jsoc channel to discuss in more detail.)

Reviewer: Ehh, no one reads to the end. It's the classic, Youtubers have to ask for subscribers at the start because that's where the viewers are. Since it's put at an interruption, it's fine. It's the opening credits 😅
